forest_json_to_csv.go

```diff
@@ -67,18 +70,194 @@ func ForestJsonToCsv(config ForestJsonToCsvConfig) error {
 	return nil
 }
 
-// TODO(hank)
 func writeForestCsv(forests []Forest, path string) error {
+	// open file
+	csvFile, err := os.Create(path)
+	if err != nil {
+		log.Fatalf("failed creating file: %s", err)
+		return err
+	}
+	defer csvFile.Close()
+	// setup columns
+	writer := csv.NewWriter(csvFile)
+	writer.Write([]string{
+		"Forest Name",
+		"Forest State",
+		"Forest URL",
+		"Forest ID",
+	})
+	// build rows
+	for _, forest := range forests {
+		rows := [][]string{}
+		rows = append(rows, append([]string{
+			forest.Name,
+			forest.State,
+			forest.Url,
+			fmt.Sprint(forest.Id),
+		}, []string{}...))
+
+		// write rows to file
+		writer.WriteAll(rows)
+	}
+	writer.Flush()
 	return nil
 }
 
-// TODO(hank)
 func writeProjectsCsv(forests []Forest, path string) error {
+	// open file
+	csvFile, err := os.Create(path)
+	if err != nil {
+		log.Fatalf("failed creating file: %s", err)
+		return err
+	}
+	defer csvFile.Close()
+
+	// build rows
+	writer := csv.NewWriter(csvFile)
+	writer.Write([]string{
+		"Forest Name",
+		"Forest State",
+		"Forest URL",
+		"Forest ID",
+		"Project Name",
+		"Project ID",
+		"Project Purposes",
+		"Project Status",
+		"Project Decision",
+		"Project Expected Implementation",
+		"Project Contact Name",
+		"Project Contact Email",
+		"Project Contact Phone",
+		"Project Description",
+		"Project Web Link",
+		"Project Location",
+		"Project Region",
+		"Project District",
+		"Project SOPA Report Date",
+		"Project Code",
+		"Project Documents",
+	})
+
+	// dedupe forests first!!
+
+	// remove all but one project update
+	forestMap := make(map[int]Forest)
+	forestIdsToRemove := []int{}
+	// inital map
+	for i, forest := range forests {
+		// check if forest in map
+		if _, ok := forestMap[forest.Id]; !ok {
+			forestMap[forest.Id] = forest
+			forestIdsToRemove = append(forestIdsToRemove, i)
+		}
+	}
+
+	// build list of leftover forests TODO clean up
+	leftoverForests := []Forest{}
+	for i, forest := range forests {
+		shouldBeRemoved := false
+		for _, index := range forestIdsToRemove {
+			if index == i {
+				shouldBeRemoved = true
+			}
+		}
+		if !shouldBeRemoved {
+			leftoverForests = append(leftoverForests, forest)
+		}
+	}
+
+	// iterate through leftovers and add states
+	for _, forest := range leftoverForests {
+		if forestInMap, ok := forestMap[forest.Id]; ok {
+			forestInMap.State = forestInMap.State + ", " + forest.State
+			forestMap[forest.Id] = forestInMap
+		}
+	}
+
```

Comment:
Maybe we make a separate `Transform` type function that captures all this business logic?

Reply:
i would also say that this code is pretty hard to follow